### PR TITLE
Created ethereum transaction and supported translation of Ride DApp metadata to ethereum ABI. Created metamask service stub.

### PR DESCRIPTION
**Pull request by esuwu**

### CODE REVIEW

File: pkg/metamask/server.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	rpc := zenrpc.NewServer(zenrpc.Options{ExposeSMD: true})
+	rpc.Register("", MetaMask{}) // public
+	rpc.Use(zenrpc.Logger(log.New(os.Stderr, "", log.LstdFlags)))
+
+	http.Handle("/", rpc)
+
+	server := &http.Server{Addr: address, Handler: nil}
+
+	go func() {
+		zap.S().Info("shutting down metamask service...")
```
Comment (nickeskov) [inline pkg/metamask/server.go:23]: You should wait for `ctx.Done`. Or you should write this line after `err := server.Shutdown(ctx)`.

File: pkg/metamask/server.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+func RunMetaMaskService(ctx context.Context, address string) error {
+	rpc := zenrpc.NewServer(zenrpc.Options{ExposeSMD: true})
+	rpc.Register("", MetaMask{}) // public
+	rpc.Use(zenrpc.Logger(log.New(os.Stderr, "", log.LstdFlags)))
+
+	http.Handle("/", rpc)
+
+	server := &http.Server{Addr: address, Handler: nil}
```
Comment (nickeskov) [inline pkg/metamask/server.go:20]: You should handle HTTP OPTIONS method.

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/metamask/server.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	rpc := zenrpc.NewServer(zenrpc.Options{ExposeSMD: true, AllowCORS: true})
+	rpc.Register("", MetaMask{}) // public
+	rpc.Use(zenrpc.Logger(log.New(os.Stderr, "", log.LstdFlags)))
+
+	http.Handle("/", rpc)
+
+	server := &http.Server{Addr: address, Handler: nil}
+
+	go func() {
+		err := server.Shutdown(ctx)
```
Comment (nickeskov) [inline pkg/metamask/server.go:23]: You should wait for `ctx.Done()`. For example:
```go
<-ctx.Done()
zap.S().Info("shutting down metamask service...")
err := server.Shutdown(ctx)
// ....
```

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: cmd/node/node.go [hunk 4]
```diff
@@ -59,6 +60,7 @@ var (
 	apiKey                                = flag.String("api-key", "", "Api key")
 	grpcAddr                              = flag.String("grpc-address", "127.0.0.1:7475", "Address for gRPC API")
 	enableGrpcApi                         = flag.Bool("enable-grpc-api", true, "Enables/disables gRPC API")
+	enableMetaMaskService                 = flag.Bool("enable-metamask", true, "Enables/disables metamask service")
```
Comment (nickeskov) [inline cmd/node/node.go:63]: I think `metaMaskServiceAddr` parameter would be useful (Address for metamask service).

File: cmd/node/node.go [hunk 5]
```diff
@@ -417,6 +419,17 @@ func main() {
 		}()
 	}
 
+	address := ":8245"
```
Comment (nickeskov) [inline cmd/node/node.go:422]: Use 8545 port as default.

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/metamask/server.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	"go.uber.org/zap"
+	"log"
+	"net/http"
+	"os"
+)
+
+func RunMetaMaskService(ctx context.Context, address string) error {
+	rpc := zenrpc.NewServer(zenrpc.Options{ExposeSMD: true, AllowCORS: true})
+	rpc.Register("", MetaMask{}) // public
+	rpc.Use(zenrpc.Logger(log.New(os.Stderr, "", log.LstdFlags)))
```
Comment (nickeskov) [inline pkg/metamask/server.go:16]: Maybe use zap logger?

File: pkg/metamask/server.go (hunk 3 above)
Comment (esuwu) [reply pkg/metamask/server.go:23]: I shouldn't. <-ctx.Done() is already in body of function server.Shutdown(ctx)

Comment (nickeskov) [reply pkg/metamask/server.go:23]: Nope, you should. In first lines in body of function `server.Shutdown(ctx)` you can see that server close listeners. It means that new incoming requests will be rejected.

File: pkg/metamask/abi/parsing_test.go [hunk 7]
```diff
@@ -40,7 +41,25 @@ func TestJsonAbi(t *testing.T) {
...
 	require.NoError(t, err)
 
-	resJson, err := getJsonAbi(callData.Signature)
+	resJson, err := getJsonAbi(callData.Signature, callData.Inputs)
+	require.NoError(t, err)
+	require.Equal(t, string(resJson), expectedJson)
+}
+
+func TestJsonAbiPayments(t *testing.T) {
+	expectedJson := `[{"name":"transfer","type":"function","inputs":[{"type":"address"},{"type":"uint256"},{"type":"[(address, uint256)]"}]}]`
```
Comment (nickeskov) [inline pkg/metamask/abi/parsing_test.go:50]: Payments ABI-JSON representation isn't correct. See [this](https://github.com/ethereum/eth-abi/issues/35) for example.

**Review by nickeskov — APPROVED** [review]

**Pull request merged**